### ISSUE TITLE
[BUG] Upgrading a 3-node cluster to 1.1.0 stuck indefinitely

### ISSUE THREAD

**Issue by janeczku** [harvester/harvester#3046]
**Describe the bug**

While upgrading a 3-node cluster from 1.0.3 to 1.1.0 the upgrade process becomes stuck indefinitely in the "pre-drain" job for the first node.

Pre-Drain Job Pod:

```
+ echo 'Waiting for volume pvc-b4ffc53c-f8d4-4c90-8011-a7f9cd156113 to be healthy...'
+ sleep 10
Waiting for volume pvc-b4ffc53c-f8d4-4c90-8011-a7f9cd156113 to be healthy...
+ '[' true ']'
+ '[' 3 -gt 2 ']'
++ kubectl get volumes.longhorn.io/pvc-b4ffc53c-f8d4-4c90-8011-a7f9cd156113 -n longhorn-system -o 'jsonpath={.status.robustness}'
+ robustness=degraded
+ '[' degraded = healthy ']'
+ '[' -f /tmp/skip-pvc-b4ffc53c-f8d4-4c90-8011-a7f9cd156113 ']'
+ echo 'Waiting for volume pvc-b4ffc53c-f8d4-4c90-8011-a7f9cd156113 to be healthy...'
+ sleep 10
Waiting for volume pvc-b4ffc53c-f8d4-4c90-8011-a7f9cd156113 to be healthy...
```

The volume is from the Prometheus deployment:

```
/home/rancher # kubectl get pv volume pvc-b4ffc53c-f8d4-4c90-8011-a7f9cd156113
NAME                                       CAPACITY   ACCESS MODES   RECLAIM POLICY   STATUS   CLAIM                                                                                                             STORAGECLASS   REASON   AGE
pvc-b4ffc53c-f8d4-4c90-8011-a7f9cd156113   50Gi       RWO            Delete           Bound    cattle-monitoring-system/prometheus-rancher-monitoring-prometheus-db-prometheus-rancher-monitoring-prometheus-0   longhorn

/home/rancher  # kubectl -n longhorn-system get lhv pvc-b4ffc53c-f8d4-4c90-8011-a7f9cd156113
NAME                    STATE   ROBUSTNESS  SCHEDULED  SIZE     NODE  AGE
pvc-b4ffc53c-f8d4-4c90-8011-a7f9cd156113  attached  degraded         53687091200  donna  169d

home/rancher # kubectl get nodes
NAME     STATUS                     ROLES                       AGE    VERSION
donna    Ready                      control-plane,etcd,master   168d   v1.22.12+rke2r1
harvey   Ready,SchedulingDisabled   control-plane,etcd,master   169d   v1.22.12+rke2r1
mike     Ready      

harvey:/home/rancher # kubectl get blockdevices.harvesterhci.io -A -o wide
NAMESPACE         NAME                               TYPE   DEVPATH    MOUNTPOINT                                                        NODENAME   STATUS   AGE
longhorn-system   502a522e036e17535bb9e7dc6cb772e6   disk   /dev/sdb   /var/lib/harvester/extra-disks/502a522e036e17535bb9e7dc6cb772e6   mike       Active   168d
longhorn-system   5c5a4447502b97530f54d41ae73ce095   disk   /dev/sdb   /var/lib/harvester/extra-disks/5c5a4447502b97530f54d41ae73ce095   harvey     Active   169d
longhorn-system   f4f0e311027560b0845129cc0f6eeb88   disk   /dev/sdb   /var/lib/harvester/extra-disks/f4f0e311027560b0845129cc0f6eeb88   donna      Active   168d
````

Longhorn Manager logs endlessly repeating:

```
time="2022-10-26T15:47:52Z" level=debug msg="Found 0 node has at least one schedulable disk" accessMode=rwo controller=longhorn-volume frontend=blockdev migratable=false node=donna owner=donna state=attached volume=pvc-b4ffc53c-f8d4-4c90-8011-a7f9cd156113
time="2022-10-26T15:47:52Z" level=error msg="There's no available disk for replica pvc-b4ffc53c-f8d4-4c90-8011-a7f9cd156113-r-afb2054f, size 53687091200"
time="2022-10-26T15:47:52Z" level=error msg="unable to schedule replica" accessMode=rwo controller=longhorn-volume frontend=blockdev migratable=false node=donna owner=donna replica=pvc-b4ffc53c-f8d4-4c90-8011-a7f9cd156113-r-afb2054f state=attached volume=pvc-b4ffc53c-f8d4-4c90-8011-a7f9cd156113
```

There is an existing issue describing the problem of the Prometheus volume never recovering: https://github.com/harvester/harvester/issues/1906

Prometheus volume is configured with 3 replicas. And since 1 of the 3 nodes is cordoned during upgrade. The volume can never recover fully blocking any further upgrade process.

**Comment by janeczku** [comment]
The LH volume resource:

```
harvey:/home/rancher # kubectl get -n longhorn-system volume pvc-b4ffc53c-f8d4-4c90-8011-a7f9cd156113 -o yaml
apiVersion: longhorn.io/v1beta2
kind: Volume
metadata:
  creationTimestamp: "2022-05-10T15:32:07Z"
  finalizers:
  - longhorn.io
  generation: 18
  labels:
    longhornvolume: pvc-b4ffc53c-f8d4-4c90-8011-a7f9cd156113
    recurring-job-group.longhorn.io/default: enabled
  name: pvc-b4ffc53c-f8d4-4c90-8011-a7f9cd156113
  namespace: longhorn-system
  resourceVersion: "224545297"
  uid: bb4ac7e7-1607-4b61-95e1-ad19d1c4da03
spec:
  Standby: false
  accessMode: rwo
  backingImage: ""
  baseImage: ""
  dataLocality: disabled
  dataSource: ""
  disableFrontend: false
  diskSelector: []
  encrypted: false
  engineImage: longhornio/longhorn-engine:v1.3.2
  fromBackup: ""
  frontend: blockdev
  lastAttachedBy: ""
  migratable: false
  migrationNodeID: ""
  nodeID: donna
  nodeSelector: []
  numberOfReplicas: 3
  recurringJobs: []
  replicaAutoBalance: ignored
  revisionCounterDisabled: false
  size: "53687091200"
  staleReplicaTimeout: 30
status:
  actualSize: 52571148288
  cloneStatus:
    snapshot: ""
    sourceVolume: ""
    state: ""
  conditions:
  - lastProbeTime: ""
    lastTransitionTime: "2022-05-10T15:32:09Z"
    message: ""
    reason: ""
    status: "False"
    type: restore
  - lastProbeTime: ""
    lastTransitionTime: "2022-10-26T15:14:21Z"
    message: replica scheduling failed
    reason: ReplicaSchedulingFailure
    status: "False"
    type: scheduled
  - lastProbeTime: ""
    lastTransitionTime: "2022-05-10T15:32:09Z"
    message: ""
    reason: ""
    status: "False"
    type: toomanysnapshots
  currentImage: longhornio/longhorn-engine:v1.3.2
  currentNodeID: donna
  expansionRequired: false
  frontendDisabled: false
  isStandby: false
  kubernetesStatus:
    lastPVCRefAt: ""
    lastPodRefAt: ""
    namespace: cattle-monitoring-system
    pvName: pvc-b4ffc53c-f8d4-4c90-8011-a7f9cd156113
    pvStatus: Bound
    pvcName: prometheus-rancher-monitoring-prometheus-db-prometheus-rancher-monitoring-prometheus-0
    workloadsStatus:
    - podName: prometheus-rancher-monitoring-prometheus-0
      podStatus: Running
      workloadName: prometheus-rancher-monitoring-prometheus
      workloadType: StatefulSet
  lastBackup: ""
  lastBackupAt: ""
  lastDegradedAt: "2022-10-26T14:57:41Z"
  ownerID: donna
  pendingNodeID: ""
  remountRequestedAt: "2022-10-26T14:56:13Z"
  restoreInitiated: false
  restoreRequired: false
  robustness: degraded
  shareEndpoint: ""
  shareState: ""
  state: attached
```

**Comment by janeczku** [comment]
Workaround:

Edit the Longhorn volume resource and change `numberOfReplicas: 3` to `numberOfReplicas: 2`

Result:

```
harvey:/home/rancher # kubectl get -n longhorn-system volume pvc-b4ffc53c-f8d4-4c90-8011-a7f9cd156113 -o wide
NAME                                       STATE      ROBUSTNESS   SCHEDULED   SIZE          NODE    AGE
pvc-b4ffc53c-f8d4-4c90-8011-a7f9cd156113   attached   healthy                  53687091200   donna   169d
```

Upgrade unblocked.

**Comment by janeczku** [comment]
You will have to repeat the workaround for two additional volumes (grafana and alertmanager pvc):
```
harvey:/home/rancher # kubectl describe volume -n longhorn-system pvc-282884ca-5453-4e0e-b1cb-640553ed70eb
Name:         pvc-282884ca-5453-4e0e-b1cb-640553ed70eb
Namespace:    longhorn-system
Labels:       longhornvolume=pvc-282884ca-5453-4e0e-b1cb-640553ed70eb
              recurring-job-group.longhorn.io/default=enabled
Annotations:  <none>
API Version:  longhorn.io/v1beta2
Kind:         Volume
Metadata:
  Creation Timestamp:  2022-05-12T09:00:19Z
  Finalizers:
    longhorn.io
  Generation:  16
  Managed Fields:
    API Version:  longhorn.io/v1beta2
    Fields Type:  FieldsV1
    fieldsV1:
      f:spec:
        f:engineImage:
        f:nodeID:
    Manager:      longhorn-manager
    Operation:    Update
    Time:         2022-10-26T14:56:10Z
    API Version:  longhorn.io/v1beta2
    Fields Type:  FieldsV1
    fieldsV1:
      f:status:
        f:actualSize:
        f:conditions:
        f:currentImage:
        f:currentNodeID:
        f:kubernetesStatus:
          f:lastPodRefAt:
          f:workloadsStatus:
        f:lastDegradedAt:
        f:ownerID:
        f:pendingNodeID:
        f:remountRequestedAt:
        f:robustness:
        f:state:
    Manager:         longhorn-manager
    Operation:       Update
    Subresource:     status
    Time:            2022-10-26T16:26:55Z
  Resource Version:  224697983
  UID:               e9c6a660-4486-49e0-82e0-68dc860e8a9b
Spec:
  Standby:           false
  Access Mode:       rwo
  Backing Image:     
  Base Image:        
  Data Locality:     disabled
  Data Source:       
  Disable Frontend:  false
  Disk Selector:
  Encrypted:          false
  Engine Image:       longhornio/longhorn-engine:v1.3.2
  From Backup:        
  Frontend:           blockdev
  Last Attached By:   
  Migratable:         false
  Migration Node ID:  
  Node ID:            mike
  Node Selector:
  Number Of Replicas:  3
  Recurring Jobs:
  Replica Auto Balance:       ignored
  Revision Counter Disabled:  false
  Size:                       10485760
  Stale Replica Timeout:      30
Status:
  Actual Size:  17608704
  Clone Status:
    Snapshot:       
    Source Volume:  
    State:          
  Conditions:
    Last Probe Time:       
    Last Transition Time:  2022-05-12T09:00:21Z
    Message:               
    Reason:                
    Status:                False
    Type:                  restore
    Last Probe Time:       
    Last Transition Time:  2022-10-26T16:26:55Z
    Message:               replica scheduling failed
    Reason:                ReplicaSchedulingFailure
    Status:                False
    Type:                  scheduled
    Last Probe Time:       
    Last Transition Time:  2022-05-12T09:00:21Z
    Message:               
    Reason:                
    Status:                False
    Type:                  toomanysnapshots
  Current Image:           longhornio/longhorn-engine:v1.3.2
  Current Node ID:         mike
  Expansion Required:      false
  Frontend Disabled:       false
  Is Standby:              false
  Kubernetes Status:
    Last PVC Ref At:  
    Last Pod Ref At:  
    Namespace:        cattle-monitoring-system
    Pv Name:          pvc-282884ca-5453-4e0e-b1cb-640553ed70eb
    Pv Status:        Bound
    Pvc Name:         rancher-monitoring-grafana
    Workloads Status:
      Pod Name:          rancher-monitoring-grafana-647fd577f4-w9vrv
      Pod Status:        Running
      Workload Name:     rancher-monitoring-grafana-647fd577f4
      Workload Type:     ReplicaSet
  Last Backup:           
  Last Backup At:        
  Last Degraded At:      2022-10-26T16:16:54Z
  Owner ID:              mike
  Pending Node ID:       
  Remount Requested At:  2022-10-26T14:55:26Z
  Restore Initiated:     false
  Restore Required:      false
  Robustness:            degraded
  Share Endpoint:        
  Share State:           
  State:                 attached
Events:                  <none>
```

**Comment by w13915984028** [comment]
cc @bk201 @starbops
<del>
@janeczku Seems 3-node is a tricky number. When ONE nodes is drained, the LH volume needs another alternative NODE to hold the traditional `3-replicas`, but the system can not meet this.  
4+ nodes cluster will have no this issue.   (another issue https://github.com/harvester/harvester/issues/2907 is solved by a related workaround)
</del>

update the suspection:

This prometheus volume pvc-b4ffc53c-f8d4-4c90-8011-a7f9cd156113 has already been in degraded status before upgrade,  the upgrade sript happens to hit this situation and wait for it to be healthy.

But considering https://github.com/harvester/harvester/issues/2833, could it be possible ?

**Comment by janeczku** [comment]
Upgrade was also blocked by `upgrade-repo-hvst-upgrade-52qzw-disk-0` volume having 3 replicas.

**Comment by janeczku** [comment]
Interestingly this issue did not exist when upgrading from 1.0.2 to 1.0.3 on same 3-node cluster @w13915984028

**Comment by w13915984028** [comment]
Technically, it is because in v1.1.0, the `upgrade_node.sh` will check `longhorn volume`, but in previous version, it does NOT. @janeczku  @bk201 

v1.0.3 (execute when v1.0.2 upgrade to v1.0.3)
https://github.com/harvester/harvester/blob/v1.0.3/package/upgrade/upgrade_node.sh#L201

v1.1.0 (execute when v1.0.3 upgrade to v1.1.0)
https://github.com/harvester/harvester/blob/master/package/upgrade/upgrade_node.sh#L303

**Comment by bk201** [comment]
@janeczku Correct. We add volume health check in this version to ensure data availability. In the previous versions, the controller just continue on next node without checking.

**Comment by bk201** [comment]
Related to https://github.com/harvester/harvester/issues/1926, we need a way to let user know what's blocking the upgrade.

**Comment by Martin-Weiss** [comment]
Beside of the health check being very useful - any idea why we can not rely on the "deprecated" status plus PDB from Longhorn?

**Comment by w13915984028** [comment]
There has already been volume status checking before upgrade. But it will only work in v1.1.0 or after.  

https://github.com/harvester/harvester/issues/2833  

The v1.0.3 version has no such checking before upgrade, that needs to be solved.

**Comment by bk201** [comment]
> #1926

#1926 is merged we can have log to debug now.
